security/authorization/engine/engine.go

```diff
@@ -0,0 +1,230 @@
+				match = false
+			} else {
+				match = out.Value().(bool)
+			}
+		}
+
+		// Process evaluation results.
+		if err != nil {
+			unknownPolicyNames = append(unknownPolicyNames, policyName)
+		} else if match {
+			return getDecision(engine, true), []string{policyName}
+		}
+	}
+	if len(unknownPolicyNames) > 0 {
+		return DecisionUnknown, unknownPolicyNames
+	}
+	return getDecision(engine, false), []string{}
+}
+
+// AuthorizationEngine is the struct for the CEL-based authorization engine.
+type AuthorizationEngine struct {
+	engines []*policyEngine
+}
+
+// NewAuthorizationEngine builds a CEL evaluation engine from a list of Envoy RBACs.
+func NewAuthorizationEngine(rbacs []*pb.RBAC) (*AuthorizationEngine, error) {
+	if len(rbacs) < 1 || len(rbacs) > 2 {
+		return &AuthorizationEngine{}, status.Errorf(codes.InvalidArgument, "must provide 1 or 2 RBACs")
+	}
+	if len(rbacs) == 2 && (rbacs[0].Action != pb.RBAC_DENY || rbacs[1].Action != pb.RBAC_ALLOW) {
+		return &AuthorizationEngine{}, status.Errorf(codes.InvalidArgument, "when providing 2 RBACs, must have 1 DENY and 1 ALLOW in that order")
+	}
+	var engines []*policyEngine
+	for _, rbac := range rbacs {
+		engines = append(engines, newPolicyEngine(rbac))
+	}
+	return &AuthorizationEngine{engines}, nil
+}
+
+// Evaluate is the core function that evaluates whether an RPC is authorized.
+//
+// ALLOW policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL-based authorization engine evaluation returns allow. If all of the RBAC
+// conditions are evaluated as false, then it returns deny. Otherwise, some
+// conditions are false and some are unknown, it returns undecided.
+//
+// DENY policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL-based authorization engine evaluation returns deny. If all of the RBAC
+// conditions are evaluated as false, then it returns allow. Otherwise, some
+// conditions are false and some are unknown, it returns undecided.
+//
+// DENY policy + ALLOW policy. Evaluation is in the following order: If one
+// of the expressions in the DENY policy is true, the authorization engine
+// returns deny. If one of the expressions in the DENY policy is unknown, it
+// returns undecided. Now all the expressions in the DENY policy are false,
+// it returns the evaluation of the ALLOW policy.
+func (authorizationEngine *AuthorizationEngine) Evaluate(args *AuthorizationArgs) (AuthorizationDecision, error) {
+	activation := args.toActivation()
+	numEngines := len(authorizationEngine.engines)
+	// If CelEvaluationEngine is used correctly, the below 2 errors should theoretically never occur.
+	// However, to ease potential debugging, they are checked anyway.
+	if numEngines < 1 || numEngines > 2 {
+		return AuthorizationDecision{}, status.Errorf(codes.Internal, "each CEL-based authorization engine should have 1 or 2 policy engines; instead, there are %d in the CEL-based authorization engine provided", numEngines)
```

Comment:
Why a gRPC status?  This is a programmer error (using `AuthorizationEngine` without using its constructor).